README.md
# Selenium Benchmark 0.2.12 (1/21/2014)
A simple way to benchmark Selenium WebDriver performance across various browsers

## Getting Started

__NOTE: You will need to set up your local test machine with the Selenium browser drivers in order to use all of the rake tasks (and the Selenium Standalone Server for Opera).__

1) `git clone https://github.com/tourdedave/selenium-benchmark.git`  
2) `bundle install`  
3) `rake server:init`  
4) `rake server:start`  
5) Open a new terminal window

`rake -T` to get a full list of commands  

```sh
terminal$ rake -T
rake benchmark:chrome[version]              # Chrome
rake benchmark:firefox[version]             # Firefox
rake benchmark:internet_explorer[version]   # IE
rake benchmark:opera[selenium_server_path]  # Opera (version 12.16)
rake report                                 # Generate report from benchmark data
rake server:init                            # Pull in the the-internet after initial checkout
rake server:start                           # Start the web server
```

6) Run your benchmarks against the browsers you want (one at a time)  

```sh
# The version number is important for reporting.
terminal$ rake benchmark:firefox['26']
terminal$ rake benchmark:internet_explorer['08']
terminal$ rake benchmark:internet_explorer['10']
terminal$ rake benchmark:chrome['32']
# Opera is the odd duck, requiring the full path to the Selenium standalone server
terminal$ rake benchmark:opera['C:\drivers\selenium-server-standalone-2.39.0.jar']
```

7) After running your benchmarks, type `rake report` to generate an aggregate data table  

![report output](https://raw2.github.com/tourdedave/selenium-benchmark/develop/benchmarks/output.png)


CHANGELOG.md
#0.2.12

+ Cleaned up the README

#0.2.11

+ Added rounding support for 0s in the data
+ Finished the instructions in the README and included an image of the report output

#0.2.10

+ Finished up rake tasks (and removed extraneous ones)
+ Added getting started instructions to the README
+ Applied a reporting fix for better handling of unsupported locator strategies in older browsers

#0.2.9

+ Improved the rounding so it will add a 0 if there are only 2 decimal places

#0.2.8

+ Trimmed the Opera table header back to '12' from '12.16'
+ Rounded the data output to 3 decimal places

#0.2.7

+ Added the ability to pass the Selenium Standalone Server path on Opera runs (which sets it as an environment variable) -- a required step to run Opera Driver

#0.2.6

+ Fixed a rendering bug with IE test data in the report

#0.2.5

+ Fixed some rendering bugs in the report generator and improved browser naming

#0.2.4

+ Took out the error output when a locator can't be found and added back in '(N/A)'

#0.2.3

+ Fixed bugs in rake tasks to launch browsers

#0.2.2

+ Added rake tasks to simplify launching each browser

#0.2.1

+ Added .gitkeep to keep the benchmark directory active in GitHub

#0.2.0
Release date: Jan 21, 2014

+ Added a Large DOM example
+ Added in report generation and made it dynamic
+ Moved some things around in anticipation of rolling things into a gem

#0.1.0
Release data: Jan 14, 2014

+ Initial benchmarking example


run.rb
require_relative 'lib/page-objects/tables'
#require_relative '../lib/page-objects/large_dom'
require_relative 'lib/table_generator'

ENV['iterations'] ||= '100'
# override the number of iterations at runtime with '`iterations=20 ruby css_vs_xpath.rb`

ENV['browser'] ||= 'firefox'
# override browser at runtime with `browser='browser_name' ruby css_vs_xpath.rb`

driver = Selenium::WebDriver.for ENV['browser'].to_sym
# Set up and start the-internet before running the test
# `cd the-internet`
# `bundle install`
# `ruby server.rb`

ENV['base_url'] = 'http://localhost:4567'

$stdout = File.new("benchmarks/#{ENV['browser']}.log", 'w')
$stdout.sync = true

tables = Tables.new(driver)
tables.visit
tables.benchmark!

#large_dom = LargeDOM.new(@driver)
#large_dom.visit
#large_Dom.benchmark!

# The benchmarking approach was borrowed from
# http://rubylearning.com/blog/2013/06/19/how-do-i-benchmark-ruby-code/

driver.quit
gen = TableGenerator.new
gen.run


lib/table_generator.rb
require 'terminal-table'

class TableGenerator

  attr_reader :logs, :harvests

  def initialize
    current_dir = File.join(File.dirname(File.expand_path(__FILE__)))
    @logs = Dir.glob(current_dir+'/../benchmarks/*.log')
    @harvests = []
    parse_logs
    prep_data
  end

  def generate
    render
  end

  private

    def parse_logs
      logs.each do |log|
        log = File.read(log)
        log.gsub(/ using \w+/,'')
        @titles  = log.scan(/^.*_\w+/).uniq
        job_log = log.scan(/real.*/m).first
        datas    = job_log.scan(/\(.*\)/).each do |data|
          data.gsub!(/\)/,'')
          data.gsub!(/\(/,'')
          data.strip!
        end
        datas.map! do |data|
          data.to_f.round(3)
        end
        datas.map! do |data|
          case data.to_s.split('.').last.size
          when 2
            data.to_s + "0"
          when 1
            data.to_s + "00"
          else
            data
          end
        end
        count = 0
        harvest = []
        @titles.count.times do |i|
          if i == 0
            harvest << [datas[i], datas[i+1]]
          else
            harvest << [datas[count], datas[count+1]]
          end
          count += 2
        end
        harvests << harvest
      end
    end

    def prep_data
      @row = []
      @titles.each {|title| @row << [title] }
      harvests.each do |harvest|
        harvest.count.times do |i|
          @row[i] << harvest[i][0]
          @row[i] << harvest[i][1]
        end
      end
    end

    def get_table_headings
      headings = ['']
      logs.each do |log|
        slug = log.scan(/benchmarks\/(.*).log/)[0][0]
        browser_name = slug.scan(/(.*)_/)[0][0]
        browser_number = slug.scan(/_(\d.*)/)[0][0]
        headings << "#{browser_name[0].capitalize} #{browser_number}"
        headings << "#{browser_name[0].capitalize} #{browser_number}"
      end
      headings.each do |heading|
        if heading.include? 'F'
          heading.gsub!(/F/,'FF')
        elsif heading.include? 'C'
          heading.gsub!(/C/,'Cr')
        elsif heading.include? 'I'
          heading.gsub!(/I/,'IE')
        elsif heading.include? 'O'
          heading.gsub!(/O/,'Op')
        else
          heading
        end
      end
      headings
    end

    def render
      headings = get_table_headings

      tmp = ['']
      ((headings.count-1)/2).times do |count|
        tmp << 'CSS'
        tmp << 'XPath'
      end
      rows = []
      rows << tmp

      @row.each do |ro|
        rows << ro
      end

      puts Terminal::Table.new headings: headings, rows: rows
    end
end


lib/page-objects/tables.rb
require_relative 'base'
require 'benchmark'

class Tables < Base

  LOCATORS = {
    :header_id_and_class => {
      :css => "#table2 thead .dues",
      :xpath => "//table[@id='table2']//thead//*[@class='dues']"
    },
    :header_id_class_and_direct_desc => {
      :css => "#table2 > thead .dues",
      :xpath => "//table[@id='table2']/thead//*[@class='dues']"
    },
    :header_traversing => {
      :css => "#table2 thead tr th:nth-of-type(4)",
      :xpath => "//table[@id='table2']//thead//tr//th[4]"
    },
    :header_traversing_and_direct_desc => {
      :css => "#table2 > thead > tr > th:nth-of-type(4)",
      :xpath => "//table[@id='table2']/thead/tr/th[4]"
    },
    :cell_id_and_class => {
      :css => "#table2 tbody .dues",
      :xpath => "//table[@id='table2']//tbody//*[@class='dues']"
    },
    :cell_id_class_and_direct_desc => {
      :css => "#table2 > tbody .dues",
      :xpath => "//table[@id='table2']/tbody//*[@class='dues']"
    },
    :cell_traversing => {
      :css => "#table2 tbody tr td:nth-of-type(4)",
      :xpath => "//table[@id='table2']//tbody//tr//td[4]"
    },
    :cell_traversing_and_direct_desc => {
      :css => "#table2 > tbody > tr > td:nth-of-type(4)",
      :xpath => "//table[@id='table2']/tbody/tr/td[4]"
    }
  }

  def visit
    go_to '/tables'
  end

  def benchmark!
    Benchmark.bmbm(27) do |bm|
      LOCATORS.each do |example, data|
        data.each do |strategy, locator|
          bm.report(example.to_s + " using " + strategy.to_s) do
            begin
              ENV['iterations'].to_i.times do
                 find(strategy => locator)
              end
            rescue Selenium::WebDriver::Error::NoSuchElementError
              puts "(N/A)"
            end
          end
        end
      end
    end
  end

end


lib/page-objects/large_dom.rb
require_relative 'base'
require 'benchmark'

class LargeDOM < Base

  LOCATORS = {
    nested_sibling_traversal: {
      css: "div#siblings > div:nth-of-type(1) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3) > div:nth-of-type(3)",
      xpath: "//div[@id='siblings']/div[1]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]/div[3]"
    },
    nested_sibling_traversal_by_class: {
      css: "div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1 > div.item-1",
      xpath: "//div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]/div[contains(@class, 'item-1')]"
    },
    table_header_id_and_class: {
      css: "table#large-table thead .column-50",
      xpath: "//table[@id='large-table']//thead//*[@class='column-50']"
    },
    table_header_id_class_and_direct_desc: {
      css: "table#large-table > thead .column-50",
      xpath: "//table[@id='large-table']/thead//*[@class='column-50']"
    },
    table_header_traversing: {
      css: "table#large-table thead tr th:nth-of-type(50)",
      xpath: "//table[@id='large-table']//thead//tr//th[50]"
    },
    table_header_traversing_and_direct_desc: {
      css: "table#large-table > thead > tr > th:nth-of-type(50)",
      xpath: "//table[@id='large-table']/thead/tr/th[50]"
    },
    table_cell_id_and_class: {
      css: "table#large-table tbody .column-50",
      xpath: "//table[@id='large-table']//tbody//*[@class='column-50']"
    },
    table_cell_id_class_and_direct_desc: {
      css: "table#large-table > tbody .column-50",
      xpath: "//table[@id='large-table']/tbody//*[@class='column-50']"
    },
    table_cell_traversing: {
      css: "table#large-table tbody tr td:nth-of-type(50)",
      xpath: "//table[@id='large-table']//tbody//tr//td[50]"
    },
    table_cell_traversing_and_direct_desc: {
      css: "table#large-table > tbody > tr > td:nth-of-type(50)",
      xpath: "//table[@id='large-table']/tbody/tr/td[50]"
    }
  }

  attr_reader :driver

  def initialize(driver)
    @driver = driver
    go_to '/large'
    is_displayed?(id: 'siblings')
    super
  end

  # The benchmarking approach was borrowed from
  # http://rubylearning.com/blog/2013/06/19/how-do-i-benchmark-ruby-code/
  def benchmark
    Benchmark.bmbm(27) do |bm|
      LOCATORS.each do |example, data|
        data.each do |strategy, locator|
          bm.report(example.to_s + " using " + strategy.to_s) do
            begin
              ENV['iterations'].to_i.times do |count|
                 find(strategy => locator)
              end
            rescue Selenium::WebDriver::Error::NoSuchElementError => error
              puts "( 0.0 )"
            end
          end
        end
      end
    end
  end

end


lib/page-objects/small_dom.rb
require_relative 'base'
require 'benchmark'

class SmallDOM < Base

  LOCATORS = {
    :header_id_and_class => {
      :css => "#table2 thead .dues",
      :xpath => "//table[@id='table2']//thead//*[@class='dues']"
    },
    :header_id_class_and_direct_desc => {
      :css => "#table2 > thead .dues",
      :xpath => "//table[@id='table2']/thead//*[@class='dues']"
    },
    :header_traversing => {
      :css => "#table2 thead tr th:nth-of-type(4)",
      :xpath => "//table[@id='table2']//thead//tr//th[4]"
    },
    :header_traversing_and_direct_desc => {
      :css => "#table2 > thead > tr > th:nth-of-type(4)",
      :xpath => "//table[@id='table2']/thead/tr/th[4]"
    },
    :cell_id_and_class => {
      :css => "#table2 tbody .dues",
      :xpath => "//table[@id='table2']//tbody//*[@class='dues']"
    },
    :cell_id_class_and_direct_desc => {
      :css => "#table2 > tbody .dues",
      :xpath => "//table[@id='table2']/tbody//*[@class='dues']"
    },
    :cell_traversing => {
      :css => "#table2 tbody tr td:nth-of-type(4)",
      :xpath => "//table[@id='table2']//tbody//tr//td[4]"
    },
    :cell_traversing_and_direct_desc => {
      :css => "#table2 > tbody > tr > td:nth-of-type(4)",
      :xpath => "//table[@id='table2']/tbody/tr/td[4]"
    }
  }

  attr_reader :driver

  def initialize(driver)
    @driver = driver
    go_to '/tables'
    super
  end

  # The benchmarking approach was borrowed from
  # http://rubylearning.com/blog/2013/06/19/how-do-i-benchmark-ruby-code/
  def benchmark
    Benchmark.bmbm(27) do |bm|
      LOCATORS.each do |example, data|
        data.each do |strategy, locator|
          bm.report(example.to_s + " using " + strategy.to_s) do
            begin
              ENV['iterations'].to_i.times do
                 find(strategy => locator)
              end
            rescue Selenium::WebDriver::Error::NoSuchElementError
              puts "( 0.0 )"
            end
          end
        end
      end
    end
  end

end


css_vs_xpath.rb
require 'rubygems'
require 'selenium-webdriver'
require 'benchmark'

LOCATORS = {
  :header_id_and_class => {
    :css => "#table2 thead .dues",
    :xpath => "//table[@id='table2']//thead//*[@class='dues']"
  },
  :header_id_class_and_direct_desc => {
    :css => "#table2 > thead .dues",
    :xpath => "//table[@id='table2']/thead//*[@class='dues']"
  },
  :header_traversing => {
    :css => "#table2 thead tr th:nth-of-type(4)",
    :xpath => "//table[@id='table2']//thead//tr//th[4]"
  },
  :header_traversing_and_direct_desc => {
    :css => "#table2 > thead > tr > th:nth-of-type(4)",
    :xpath => "//table[@id='table2']/thead/tr/th[4]"
  },
  :cell_id_and_class => {
    :css => "#table2 tbody .dues",
    :xpath => "//table[@id='table2']//tbody//*[@class='dues']"
  },
  :cell_id_class_and_direct_desc => {
    :css => "#table2 > tbody .dues",
    :xpath => "//table[@id='table2']/tbody//*[@class='dues']"
  },
  :cell_traversing => {
    :css => "#table2 tbody tr td:nth-of-type(4)",
    :xpath => "//table[@id='table2']//tbody//tr//td[4]"
  },
  :cell_traversing_and_direct_desc => {
    :css => "#table2 > tbody > tr > td:nth-of-type(4)",
    :xpath => "//table[@id='table2']/tbody/tr/td[4]"
  }
}

ENV['iterations'] ||= '100'
# override the number of iterations at runtime with '`iterations=20 ruby css_vs_xpath.rb`

ENV['browser'] ||= 'firefox'
# override browser at runtime with `browser='browser_name' ruby css_vs_xpath.rb`

driver = Selenium::WebDriver.for ENV['browser'].to_sym

# Set up and start the-internet before running the test
# `cd the-internet`
# `bundle install`
# `ruby server.rb`

$stdout = File.new("benchmark_#{ENV['browser']}.log", 'w')
$stdout.sync = true

driver.get "http://localhost:4567/tables"

# The benchmarking approach was borrowed from
# http://rubylearning.com/blog/2013/06/19/how-do-i-benchmark-ruby-code/
Benchmark.bmbm(27) do |bm|
  LOCATORS.each do |example, data|
    data.each do |strategy, locator|
      bm.report(example.to_s + " using " + strategy.to_s) do
        begin
          ENV['iterations'].to_i.times do
             driver.find_element(strategy => locator)
          end
        rescue Selenium::WebDriver::Error::NoSuchElementError
          puts "N/A"
        end
      end
    end
  end
end

driver.quit
